src/Interpreters/Context.cpp

```diff
@@ -528,7 +546,19 @@ void Context::copyFrom(const ContextPtr & other)
     *this = *other;
 }
 
-Context::~Context() = default;
+Context::~Context()
+{
+    if (user_id && *user_id != UUIDHelpers::Nil && this == this->session_context.lock().get())
+    {
+        auto lock = getLock();
+        if (auto session_log = getSessionLog())
+        {
+            const auto user = (access ? access->getUser() : UserPtr{});
+            if (user)
+                session_log->add(makeSessionLogElement(SESSION_LOGOUT, user, client_info));
+        }
```

Comment:
Custom code in destructors rarely looks good. I think we can add a new member `ext::scope_guard add_logout_event_to_session_log` to the `Context` class, and set it in `setUser()` right after you add a login event to `system.session_log`. It allows to keep the destructor default, keep all `session_log`-related code in one place, and ensure that we add a logout event only if we add a login event.

Reply:
unfortunately, `ext::scope_guard` would make `Context` uncopyable and I would have to craft a custom full-blown constructor.